### PR TITLE
Invalidate cache for link[preload] in DEV

### LINKED ISSUE #5860: Safari cache on dev
**The context**
Safari 12.0.1
nextjs in dev mode
/pages/index.js with simplest react code:

```
import React from 'react'

class PageIndex extends React.Component{
    constructor(props){
        super(props)
    }
    render(){
        return (<div>123</div>)
    }
}
export default PageIndex;
```

**Behaviour**
1. I change text to 1234
2. Change blinks to 1234 and gets back to 123
3. In network I see "in memory" near index.js
4. This little icon in network, develop->empty caches, option + cmd + R, alltogether has no impact on this fact

### PR DESCRIPTION
**Pull request by giuseppeg**
Fixes #5860

### CODE REVIEW

**Review by timneutkens — CHANGES_REQUESTED** [review]
Actually only adding it to preload defeats the purpose of preload right 🤔 as it would be 2 unique requests

**Comment by giuseppeg** [comment]
> Actually only adding it to preload defeats the purpose of preload right 🤔 as it would be 2 unique requests

yes I think so, I'll try to get the npm link-ing work later and try it out.

**Comment by giuseppeg** [comment]
@timneutkens would it make sense to map over the head and body children, look for links with preload tags and then add the timestamp to both them and the matching scrip tags? Or do you prefer the hardcoded version?

**Comment by giuseppeg** [comment]
alright it should be better now :)

**Comment by timneutkens** [comment]
@giuseppeg I actually think you could add the timestamp on the renderer level, next-server/server/render.tsx, as then you don't even have to manage adding it in multiple places 👍

**Comment by giuseppeg** [comment]
oh yeah definitely! Sorry I am not that familiar with the codebase yet 😅

**Comment by timneutkens** [comment]
@giuseppeg no worries 🙌 This would actually be the correct solution on older versions of Next.js, that bundle rendering was added in v6

File: packages/next/pages/_document.js [hunk 1]
```diff
@@ -119,13 +123,12 @@ export class Head extends Component {
...
-      <link rel='preload' href={`${assetPrefix}/_next/static/${buildId}/pages/_app.js`} as='script' nonce={this.props.nonce} crossOrigin={this.props.crossOrigin || process.crossOrigin} />
-      <link rel='preload' href={`${assetPrefix}/_next/static/${buildId}/pages/_error.js`} as='script' nonce={this.props.nonce} crossOrigin={this.props.crossOrigin || process.crossOrigin} />
+      {page !== '/_error' && <link rel='preload' href={`${assetPrefix}/_next/static/${buildId}/pages${pagePathname}${this.preloadLinkTagInvalidate}`} as='script' nonce={this.props.nonce} crossOrigin={this.props.crossOrigin || process.crossOrigin} />}
+      <link rel='preload' href={`${assetPrefix}/_next/static/${buildId}/pages/_app.js${this.preloadLinkTagInvalidate}`} as='script' nonce={this.props.nonce} crossOrigin={this.props.crossOrigin || process.crossOrigin} />
+      <link rel='preload' href={`${assetPrefix}/_next/static/${buildId}/pages/_error.js${this.preloadLinkTagInvalidate}`} as='script' nonce={this.props.nonce} crossOrigin={this.props.crossOrigin || process.crossOrigin} />
       {this.getPreloadDynamicChunks()}
       {this.getPreloadMainLinks()}
```
[the remark this answers is not in the document]
Comment (timneutkens) [reply packages/next/pages/_document.js]: The dynamic ones here are actually needed for hydration anyway, so the other isn't important as we'd wait for these to load before doing anything else anyway currently

[the remark this answers is not in the document]
Comment (timneutkens) [reply packages/next/pages/_document.js]: Note these are only the dynamic components that *have been rendered* not *all*

[the remark this answers is not in the document]
Comment (giuseppeg) [reply packages/next/pages/_document.js]: ah sure, is that the case for client side rendering too though?

[the remark this answers is not in the document]
Comment (timneutkens) [reply packages/next/pages/_document.js]: @giuseppeg those are handled by webpack's dynamic loading mechanism.

File: packages/next/pages/_document.js [hunk 2]
```diff
@@ -16,7 +16,7 @@ export default class Document extends Component {
   static getInitialProps ({ renderPage }) {
     const { html, head } = renderPage()
     const styles = flush()
-    return { html, head, styles }
+    return { html, head, styles, timestamp: Date.now() }
```
Comment (timneutkens) [inline packages/next/pages/_document.js]: Can we disable this in prod, probably not too heavy but still not useful to have `Date.now` running on every request.

Comment (giuseppeg) [reply packages/next/pages/_document.js]: totally

File: test/integration/app-document/test/rendering.js [hunk 3]
```diff
@@ -67,6 +67,18 @@ export default function ({ app }, suiteName, render, fetch) {
         expect($('#render-page-enhance-app').text().includes(nonce)).toBe(true)
         expect($('#render-page-enhance-component').text().includes(nonce)).toBe(true)
       })
+
+      // This is a workaround to fix https://github.com/zeit/next.js/issues/5860
+      // TODO: remove this workaround when https://bugs.webkit.org/show_bug.cgi?id=187726 is fixed.
+      test('It adds a timestamp to link tags with preload attribute to invalidate the cache (DEV only)', async () => {
```
Comment (timneutkens) [inline test/integration/app-document/test/rendering.js]: Could you also add a test to `integration/production` to make sure we don't regress there in the future.

File: packages/next/pages/_document.js [hunk 4]
```diff
@@ -119,13 +123,12 @@ export class Head extends Component {
...
     }
-
     return <head {...this.props}>
       {children}
       {head}
-      {page !== '/_error' && <link rel='preload' href={`${assetPrefix}/_next/static/${buildId}/pages${pagePathname}`} as='script' nonce={this.props.nonce} crossOrigin={this.props.crossOrigin || process.crossOrigin} />}
-      <link rel='preload' href={`${assetPrefix}/_next/static/${buildId}/pages/_app.js`} as='script' nonce={this.props.nonce} crossOrigin={this.props.crossOrigin || process.crossOrigin} />
-      <link rel='preload' href={`${assetPrefix}/_next/static/${buildId}/pages/_error.js`} as='script' nonce={this.props.nonce} crossOrigin={this.props.crossOrigin || process.crossOrigin} />
+      {page !== '/_error' && <link rel='preload' href={`${assetPrefix}/_next/static/${buildId}/pages${pagePathname}${this.preloadLinkTagInvalidate}`} as='script' nonce={this.props.nonce} crossOrigin={this.props.crossOrigin || process.crossOrigin} />}
+      <link rel='preload' href={`${assetPrefix}/_next/static/${buildId}/pages/_app.js${this.preloadLinkTagInvalidate}`} as='script' nonce={this.props.nonce} crossOrigin={this.props.crossOrigin || process.crossOrigin} />
```
[the remark this answers is not in the document]
Comment (giuseppeg) [reply packages/next/pages/_document.js]: @timneutkens what is your preference here? Should we stick to the current revision or should I move (hardcode) these files in `next-server/server/render.tsx` and go for your suggestion? There are pro and cons with both but I don't have a strong preference.

[the remark this answers is not in the document]
Comment (timneutkens) [reply packages/next/pages/_document.js]: Let's keep it like this

File: packages/next/pages/_document.js [hunk 5]
```diff
@@ -16,7 +16,15 @@ export default class Document extends Component {
   static getInitialProps ({ renderPage }) {
     const { html, head } = renderPage()
     const styles = flush()
-    return { html, head, styles }
+    const initialProps = { html, head, styles }
+
+    // This is a workaround to fix https://github.com/zeit/next.js/issues/5860
+    // TODO: remove this workaround when https://bugs.webkit.org/show_bug.cgi?id=187726 is fixed.
+    if (process.env.NODE_ENV !== 'production') {
+      initialProps.timestamp = Date.now()
```
Comment (giuseppeg) [inline packages/next/pages/_document.js]: a type checker would yell at me so badly here 😅

File: packages/next/pages/_document.js [hunk 6]
```diff
@@ -10,7 +10,8 @@ const Fragment = React.Fragment || function Fragment ({ children }) {
 
 export default class Document extends Component {
   static childContextTypes = {
-    _documentProps: PropTypes.any
+    _documentProps: PropTypes.any,
+    _devOnlyInvalidateCacheQueryString: PropTypes.string,
```
Comment (timneutkens) [inline packages/next/pages/_document.js]: Should this use 16.3 context instead?

Comment (timneutkens) [reply packages/next/pages/_document.js]: The documentProps one is kinda tricky but your new one might make sense

**Review by timneutkens — APPROVED** [review]

File: packages/next/pages/_document.js (hunk 6 above)
Comment (giuseppeg) [reply packages/next/pages/_document.js]: I will think of some way to refactor this. With the current implementation I think that we would need to add the provider in next-server as the Document `render` method could be overridden. 

After all probably your idea of adding the timestamp in next-server was better - we could define all the files to render there (also the hardcoded ones) and have this component just render them.

**Pull request merged**